crates/iceberg/src/spec/arrow.rs

```diff
@@ -0,0 +1,553 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use crate::error::Result;
+use crate::spec::{
+    ListType, MapType, NestedField, NestedFieldRef, PrimitiveType, Schema, StructType, Type,
+};
+use crate::{Error, ErrorKind};
+use arrow_schema::{DataType, Field, Fields, Schema as ArrowSchema, TimeUnit};
+use std::sync::Arc;
+
+/// A post order arrow schema visitor.
+///
+/// For order of methods called, please refer to [`visit_schema`].
+pub trait ArrowSchemaVisitor {
+    /// Return type of this visitor.
+    type T;
+    type U;
+
+    /// Called before struct/list/map field.
+    fn before_field(&mut self, _field: &Field) -> Result<()> {
+        Ok(())
+    }
+
+    /// Called after struct/list/map field.
+    fn after_field(&mut self, _field: &Field) -> Result<()> {
+        Ok(())
+    }
+
+    /// Called after schema's type visited.
+    fn schema(&mut self, schema: &ArrowSchema, values: Vec<Self::T>) -> Result<Self::U>;
+
+    /// Called after struct's fields visited.
+    fn r#struct(&mut self, fields: &Fields, results: Vec<Self::T>) -> Result<Self::T>;
+
+    /// Called after list fields visited.
+    fn list(&mut self, list: &DataType, value: Self::T) -> Result<Self::T>;
+
+    /// Called after map's key and value fields visited.
+    fn map(&mut self, map: &DataType, key_value: Self::T, value: Self::T) -> Result<Self::T>;
+
+    /// Called when see a primitive type.
+    fn primitive(&mut self, p: &DataType) -> Result<Self::T>;
+}
+
+/// Visiting a type in post order.
+fn visit_type<V: ArrowSchemaVisitor>(r#type: &DataType, visitor: &mut V) -> Result<V::T> {
+    match r#type {
+        p if p.is_primitive()
+            || matches!(
+                p,
+                DataType::Boolean
+                    | DataType::Utf8
+                    | DataType::Binary
+                    | DataType::FixedSizeBinary(_)
+            ) =>
+        {
+            visitor.primitive(p)
+        }
+        DataType::List(element_field) => {
+            visitor.before_field(element_field)?;
```

Comment:
I think we need to distinguish `field` for `list_element`, one use case is the full name of field, for list the field name is overrided, see https://github.com/apache/iceberg-python/blob/afdfa351119090f09d38ef72857d6303e691f5ad/pyiceberg/io/pyarrow.py#L812